src/software/geom/algorithms/step_along_perimeter.cpp

```diff
@@ -0,0 +1,75 @@
+#include "software/geom/algorithms/step_along_perimeter.h"
+
+#include <algorithm>
+#include <vector>
+
+#include "software/geom/algorithms/collinear.h"
+#include "software/geom/algorithms/distance.h"
+#include "software/geom/segment.h"
+
+
+Point stepAlongPerimeter(const Polygon& polygon, const Point& start,
+                         double travel_distance)
+{
+    if (travel_distance == 0.0)
+    {
+        return start;
+    }
+
+    std::vector<Segment> polygon_segments = polygon.getSegments();
+    std::size_t start_segment_index       = 0;
+
+    // find initial segment which contains start point
+    auto it =
+        std::find_if(polygon_segments.begin(), polygon_segments.end(),
+                     [&start](const Segment& segment) {
+                         return collinear(segment.getStart(), start, segment.getEnd());
+                     });
+
+    start_segment_index = std::distance(polygon_segments.begin(), it);
```

Comment:
Curious what happens if the point is not on the perimeter? Would we get segfault due to the array out of bound access on line 44?
If that's the case, and we definitely want to keep this requirement, then we should add a `CHECK` here (basically an assert by g3log library) with a error msg so it is easier to track down the crash. Though ideally we should not crash.